File: articles/behavior-genetics-and-the-biological-basis-of-personality/python/analyze_behavior_genetics_personality.py
#!/usr/bin/env python3
from pathlib import Path
import json
import numpy as np
import pandas as pd

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "data" / "synthetic_personality_twin_data.csv"
OUT = ROOT / "outputs"
OUT.mkdir(exist_ok=True)

REQUIRED = {
    "pair_id",
    "zygosity",
    "genetic_relatedness",
    "twin1_trait",
    "twin2_trait",
    "twin1_temperament_reactivity",
    "twin2_temperament_reactivity",
    "twin1_effortful_control",
    "twin2_effortful_control",
    "family_stress",
    "social_support",
    "socioeconomic_security",
    "educational_stability",
    "shared_environment_index",
    "nonshared_environment_index",
    "trait_mean",
    "trait_difference",
    "gxe_marker",
    "rge_marker",
    "developmental_context_score",
}

def twin_corr(frame: pd.DataFrame, col1: str, col2: str) -> float:
    return float(frame[[col1, col2]].corr().iloc[0, 1])

def rough_ace(r_mz: float, r_dz: float) -> dict:
    return {
        "additive_genetic_h2": 2 * (r_mz - r_dz),
        "shared_environment_c2": 2 * r_dz - r_mz,
        "nonshared_environment_e2": 1 - r_mz,
    }

def ols(df: pd.DataFrame, y: str, xs: list[str]) -> dict:
    model_df = df[[y] + xs].dropna()
    X = np.column_stack([np.ones(len(model_df)), model_df[xs].to_numpy(float)])
    beta, *_ = np.linalg.lstsq(X, model_df[y].to_numpy(float), rcond=None)
    pred = X @ beta
    ss_res = float(np.sum((model_df[y].to_numpy(float) - pred) ** 2))
    ss_tot = float(np.sum((model_df[y].to_numpy(float) - model_df[y].mean()) ** 2))
    r2 = 1.0 - ss_res / ss_tot if ss_tot else float("nan")
    return {
        "coefficients": dict(zip(["intercept", *xs], map(float, beta))),
        "r_squared": r2,
        "n": int(len(model_df)),
    }

df = pd.read_csv(DATA)
missing = REQUIRED - set(df.columns)
if missing:
    raise ValueError(f"Missing columns: {sorted(missing)}")

cor_rows = []
for outcome, c1, c2 in [
    ("personality_trait", "twin1_trait", "twin2_trait"),
    ("temperament_reactivity", "twin1_temperament_reactivity", "twin2_temperament_reactivity"),
    ("effortful_control", "twin1_effortful_control", "twin2_effortful_control"),
]:
    for zygosity, group in df.groupby("zygosity"):
        cor_rows.append({
            "outcome": outcome,
            "zygosity": zygosity,
            "n_pairs": int(len(group)),
            "twin_correlation": twin_corr(group, c1, c2),
        })

correlations = pd.DataFrame(cor_rows)

ace_rows = []
for outcome in correlations["outcome"].unique():
    sub = correlations[correlations["outcome"] == outcome]
    r_mz = float(sub.loc[sub["zygosity"] == "MZ", "twin_correlation"].iloc[0])
    r_dz = float(sub.loc[sub["zygosity"] == "DZ", "twin_correlation"].iloc[0])
    estimates = rough_ace(r_mz, r_dz)
    for component, estimate in estimates.items():
        ace_rows.append({
            "outcome": outcome,
            "component": component,
            "estimate": estimate,
            "r_mz": r_mz,
            "r_dz": r_dz,
        })

ace_summary = pd.DataFrame(ace_rows)

rng = np.random.default_rng(20260522)
boot_rows = []
for replicate in range(500):
    sampled = []
    for zygosity, group in df.groupby("zygosity"):
        sampled.append(group.sample(n=len(group), replace=True, random_state=int(rng.integers(0, 1_000_000))))
    boot = pd.concat(sampled, ignore_index=True)
    r_mz = twin_corr(boot[boot["zygosity"] == "MZ"], "twin1_trait", "twin2_trait")
    r_dz = twin_corr(boot[boot["zygosity"] == "DZ"], "twin1_trait", "twin2_trait")
    est = rough_ace(r_mz, r_dz)
    boot_rows.append({"replicate": replicate + 1, **est})

bootstrap = pd.DataFrame(boot_rows)

bootstrap_summary = pd.DataFrame([
    {
        "component": component,
        "mean": float(bootstrap[component].mean()),
        "lower_95": float(bootstrap[component].quantile(0.025)),
        "upper_95": float(bootstrap[component].quantile(0.975)),
    }
    for component in ["additive_genetic_h2", "shared_environment_c2", "nonshared_environment_e2"]
])

pair_summary = (
    df.groupby("zygosity")
    .agg(
        n_pairs=("pair_id", "count"),
        trait_mean=("trait_mean", "mean"),
        trait_difference_mean=("trait_difference", "mean"),
        family_stress_mean=("family_stress", "mean"),
        social_support_mean=("social_support", "mean"),
        socioeconomic_security_mean=("socioeconomic_security", "mean"),
        nonshared_environment_mean=("nonshared_environment_index", "mean"),
        gxe_marker_mean=("gxe_marker", "mean"),
        rge_marker_mean=("rge_marker", "mean"),
        developmental_context_mean=("developmental_context_score", "mean"),
    )
    .reset_index()
)

df["zygosity_mz"] = (df["zygosity"] == "MZ").astype(int)
for variable in ["family_stress", "social_support", "socioeconomic_security", "educational_stability"]:
    df[f"{variable}_centered"] = df[variable] - df[variable].mean()

models = {
    "trait_difference_environmental_moderation": ols(
        df,
        "trait_difference",
        ["zygosity_mz", "family_stress_centered", "social_support_centered", "socioeconomic_security_centered", "nonshared_environment_index"],
    ),
    "gxe_marker_context": ols(
        df,
        "gxe_marker",
        ["genetic_relatedness", "family_stress_centered", "social_support_centered", "educational_stability_centered"],
    ),
    "rge_marker_context": ols(
        df,
        "rge_marker",
        ["trait_mean", "social_support_centered", "socioeconomic_security_centered", "developmental_context_score"],
    ),
    "professional_use_boundary": "Suitable for education, research prototyping, consulting support, organizational learning, and methodological demonstration; not a standalone genetic assessment, screening, prediction, or decision system.",
}

correlations.to_csv(OUT / "python_twin_correlations.csv", index=False)
ace_summary.to_csv(OUT / "python_rough_ace_summary.csv", index=False)
bootstrap.to_csv(OUT / "python_ace_bootstrap.csv", index=False)
bootstrap_summary.to_csv(OUT / "python_ace_bootstrap_summary.csv", index=False)
pair_summary.to_csv(OUT / "python_pair_summary.csv", index=False)
df.to_csv(OUT / "python_scored_personality_twin_data.csv", index=False)
df[
    [
        "genetic_relatedness",
        "twin1_trait",
        "twin2_trait",
        "trait_mean",
        "trait_difference",
        "family_stress",
        "social_support",
        "socioeconomic_security",
        "educational_stability",
        "shared_environment_index",
        "nonshared_environment_index",
        "gxe_marker",
        "rge_marker",
        "developmental_context_score",
    ]
].corr().to_csv(OUT / "python_behavior_genetics_correlations.csv")
(OUT / "python_models.json").write_text(json.dumps(models, indent=2))
print(f"Wrote outputs to {OUT}")

File: articles/behavior-genetics-and-the-biological-basis-of-personality/r/analyze_behavior_genetics_personality.R
root <- normalizePath(file.path(dirname(sys.frame(1)$ofile), ".."))
data_path <- file.path(root, "data", "synthetic_personality_twin_data.csv")
outputs <- file.path(root, "outputs")
dir.create(outputs, showWarnings = FALSE, recursive = TRUE)

data <- read.csv(data_path, stringsAsFactors = FALSE)

twin_corr <- function(df, col1, col2) {
  cor(df[[col1]], df[[col2]], use = "pairwise.complete.obs")
}

rough_ace <- function(r_mz, r_dz) {
  data.frame(
    component = c("additive_genetic_h2", "shared_environment_c2", "nonshared_environment_e2"),
    estimate = c(2 * (r_mz - r_dz), 2 * r_dz - r_mz, 1 - r_mz)
  )
}

outcomes <- list(
  personality_trait = c("twin1_trait", "twin2_trait"),
  temperament_reactivity = c("twin1_temperament_reactivity", "twin2_temperament_reactivity"),
  effortful_control = c("twin1_effortful_control", "twin2_effortful_control")
)

correlation_rows <- data.frame()
ace_rows <- data.frame()

for (name in names(outcomes)) {
  cols <- outcomes[[name]]
  mz <- data[data$zygosity == "MZ", ]
  dz <- data[data$zygosity == "DZ", ]
  r_mz <- twin_corr(mz, cols[1], cols[2])
  r_dz <- twin_corr(dz, cols[1], cols[2])

  correlation_rows <- rbind(
    correlation_rows,
    data.frame(outcome = name, zygosity = "MZ", n_pairs = nrow(mz), twin_correlation = r_mz),
    data.frame(outcome = name, zygosity = "DZ", n_pairs = nrow(dz), twin_correlation = r_dz)
  )

  ace <- rough_ace(r_mz, r_dz)
  ace$outcome <- name
  ace$r_mz <- r_mz
  ace$r_dz <- r_dz
  ace_rows <- rbind(ace_rows, ace)
}

set.seed(20260522)
boot_rows <- data.frame()
for (i in 1:500) {
  mz_sample <- data[data$zygosity == "MZ", ][sample(which(data$zygosity == "MZ"), replace = TRUE), ]
  dz_sample <- data[data$zygosity == "DZ", ][sample(which(data$zygosity == "DZ"), replace = TRUE), ]

  r_mz <- twin_corr(mz_sample, "twin1_trait", "twin2_trait")
  r_dz <- twin_corr(dz_sample, "twin1_trait", "twin2_trait")
  ace <- rough_ace(r_mz, r_dz)

  boot_rows <- rbind(
    boot_rows,
    data.frame(
      replicate = i,
      additive_genetic_h2 = ace$estimate[ace$component == "additive_genetic_h2"],
      shared_environment_c2 = ace$estimate[ace$component == "shared_environment_c2"],
      nonshared_environment_e2 = ace$estimate[ace$component == "nonshared_environment_e2"]
    )
  )
}

boot_summary <- data.frame(
  component = c("additive_genetic_h2", "shared_environment_c2", "nonshared_environment_e2"),
  mean = c(mean(boot_rows$additive_genetic_h2), mean(boot_rows$shared_environment_c2), mean(boot_rows$nonshared_environment_e2)),
  lower_95 = c(quantile(boot_rows$additive_genetic_h2, 0.025), quantile(boot_rows$shared_environment_c2, 0.025), quantile(boot_rows$nonshared_environment_e2, 0.025)),
  upper_95 = c(quantile(boot_rows$additive_genetic_h2, 0.975), quantile(boot_rows$shared_environment_c2, 0.975), quantile(boot_rows$nonshared_environment_e2, 0.975))
)

data$zygosity_mz <- ifelse(data$zygosity == "MZ", 1, 0)
data$family_stress_centered <- data$family_stress - mean(data$family_stress)
data$social_support_centered <- data$social_support - mean(data$social_support)
data$socioeconomic_security_centered <- data$socioeconomic_security - mean(data$socioeconomic_security)
data$educational_stability_centered <- data$educational_stability - mean(data$educational_stability)

model_difference <- lm(
  trait_difference ~ zygosity_mz + family_stress_centered +
    social_support_centered + socioeconomic_security_centered +
    nonshared_environment_index,
  data = data
)

model_gxe <- lm(
  gxe_marker ~ genetic_relatedness + family_stress_centered +
    social_support_centered + educational_stability_centered,
  data = data
)

model_rge <- lm(
  rge_marker ~ trait_mean + social_support_centered +
    socioeconomic_security_centered + developmental_context_score,
  data = data
)

pair_summary <- aggregate(
  data[, c("trait_mean", "trait_difference", "family_stress", "social_support", "socioeconomic_security", "nonshared_environment_index", "gxe_marker", "rge_marker", "developmental_context_score")],
  by = list(zygosity = data$zygosity),
  FUN = mean
)

corr_cols <- c(
  "genetic_relatedness",
  "twin1_trait",
  "twin2_trait",
  "trait_mean",
  "trait_difference",
  "family_stress",
  "social_support",
  "socioeconomic_security",
  "educational_stability",
  "shared_environment_index",
  "nonshared_environment_index",
  "gxe_marker",
  "rge_marker",
  "developmental_context_score"
)

write.csv(correlation_rows, file.path(outputs, "r_twin_correlations.csv"), row.names = FALSE)
write.csv(ace_rows, file.path(outputs, "r_rough_ace_summary.csv"), row.names = FALSE)
write.csv(boot_rows, file.path(outputs, "r_ace_bootstrap.csv"), row.names = FALSE)
write.csv(boot_summary, file.path(outputs, "r_ace_bootstrap_summary.csv"), row.names = FALSE)
write.csv(pair_summary, file.path(outputs, "r_pair_summary.csv"), row.names = FALSE)
write.csv(cor(data[, corr_cols], use = "pairwise.complete.obs"), file.path(outputs, "r_behavior_genetics_correlations.csv"))
capture.output(summary(model_difference), file = file.path(outputs, "r_trait_difference_model_summary.txt"))
capture.output(summary(model_gxe), file = file.path(outputs, "r_gxe_marker_model_summary.txt"))
capture.output(summary(model_rge), file = file.path(outputs, "r_rge_marker_model_summary.txt"))
write.csv(data, file.path(outputs, "r_scored_personality_twin_data.csv"), row.names = FALSE)
cat("Wrote R outputs to:", outputs, "\n")
